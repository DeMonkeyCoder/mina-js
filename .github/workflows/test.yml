name: Run tests
on:
    push:
        branches:
            - main
jobs:
    my-job:
        name: my-job
        runs-on: ubuntu-latest
        steps:
            - uses: actions/checkout@v4
            - uses: oven-sh/setup-bun@v2
            - run: bun install
            - run: bun run build
            - run: bun run test
