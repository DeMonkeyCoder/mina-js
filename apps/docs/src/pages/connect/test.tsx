import { TestZkApp } from "../../components/test-zkapp";

const TestPage = () => {
	if (import.meta.env.SSR) return null;
	return <TestZkApp />;
};

export default TestPage;
