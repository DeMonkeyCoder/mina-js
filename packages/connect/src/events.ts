import type {
	MinaAnnounceProviderEvent,
	MinaProviderDetail,
} from "@mina-js/providers";

export type AnnounceProviderReturnType = () => void;

/**
 * Creates an event to announce a Mina provider and registers a handler to respond to subsequent requests.
 *
 * @param {MinaProviderDetail} detail - The details of the provider to announce.
 * @returns {AnnounceProviderReturnType} A function to remove the event listener when it is no longer needed.
 */
export function announceProvider(
	detail: MinaProviderDetail,
): AnnounceProviderReturnType {
	const event: CustomEvent<MinaProviderDetail> = new CustomEvent(
		"mina:announceProvider",
		{ detail: Object.freeze(detail) },
	);
	window.dispatchEvent(event);
	const handler = () => window.dispatchEvent(event);
	window.addEventListener("mina:requestProvider", handler);
	return () => window.removeEventListener("mina:requestProvider", handler);
}

export type RequestProvidersParameters = (
	providerDetail: MinaProviderDetail,
) => void;
export type RequestProvidersReturnType = (() => void) | undefined;

/**
 * Requests providers to be announced.
 *
 * This function adds an event listener for "mina:announceProvider" events and
 * dispatches a custom event named "mina:requestProvider" to trigger the announce.
 *
 * @param listener A callback function that will be called when a provider is announced.
 */
export function requestProviders(
	listener: RequestProvidersParameters,
): RequestProvidersReturnType {
	if (typeof window === "undefined") return;
	const handler = (event: MinaAnnounceProviderEvent) => listener(event.detail);

	window.addEventListener("mina:announceProvider", handler as never);

	window.dispatchEvent(new CustomEvent("mina:requestProvider"));

	return () =>
		window.removeEventListener("mina:announceProvider", handler as never);
}
