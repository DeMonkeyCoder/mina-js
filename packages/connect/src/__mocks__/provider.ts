import { mock } from "bun:test";
import type { MinaProviderDetail } from "@mina-js/providers";

export const mockedProvider: MinaProviderDetail = {
	info: {
		name: "Pallad",
		icon: "data:image/",
		rdns: "co.pallad",
		slug: "pallad",
	},
	provider: {
		request: mock(),
		addListener: mock(),
		removeListener: mock(),
	},
};
