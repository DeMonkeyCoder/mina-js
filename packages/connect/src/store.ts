import type { MinaProviderDetail } from "@mina-js/providers";
import { requestProviders } from "./events";

export type Listener = (providerDetails: readonly MinaProviderDetail[]) => void;

export type Store = {
	/**
	 * Clears the store, including all provider details.
	 */
	clear(): void;
	/**
	 * Destroys the store, including all provider details and listeners.
	 */
	destroy(): void;
	/**
	 * Finds a provider detail by its slug.
	 */
	findProvider(args: { slug: string }): MinaProviderDetail | undefined;
	/**
	 * Returns all provider details that have been emitted.
	 */
	getProviders(): readonly MinaProviderDetail[];
	/**
	 * Resets the store, and emits an event to request provider details.
	 */
	reset(): void;
	/**
	 * Subscribes to emitted provider details.
	 */
	subscribe(
		listener: Listener,
		args?: { emitImmediately?: boolean | undefined } | undefined,
	): () => void;

	/**
	 * @internal
	 * Current state of listening listeners.
	 */
	_listeners(): Set<Listener>;
};

export function createStore(): Store {
	const listeners: Set<Listener> = new Set();
	let providerDetails: readonly MinaProviderDetail[] = [];

	const request = () =>
		requestProviders((providerDetail) => {
			if (
				providerDetails.some(
					({ info }) => info.slug === providerDetail.info.slug,
				)
			)
				return;

			providerDetails = [...providerDetails, providerDetail];
			for (const listener of listeners) {
				listener(providerDetails);
			}
		});
	let unwatch = request();

	return {
		_listeners() {
			return listeners;
		},
		clear() {
			for (const listener of listeners) {
				listener([]);
			}
			providerDetails = [];
		},
		destroy() {
			this.clear();
			listeners.clear();
			unwatch?.();
		},
		findProvider({ slug }) {
			return providerDetails.find(
				(providerDetail) => providerDetail.info.slug === slug,
			);
		},
		getProviders() {
			return providerDetails;
		},
		reset() {
			this.clear();
			unwatch?.();
			unwatch = request();
		},
		subscribe(listener, { emitImmediately } = {}) {
			listeners.add(listener);
			if (emitImmediately) listener(providerDetails);
			return () => listeners.delete(listener);
		},
	};
}
