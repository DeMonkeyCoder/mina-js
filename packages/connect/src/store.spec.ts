import { expect, it, mock } from "bun:test";
import { mockedProvider } from "./__mocks__/provider";
import { announceProvider } from "./events";
import { createStore } from "./store";

it("should initialize the store", () => {
	const store = createStore();
	expect(store).toMatchSnapshot();
});

it("should update providers", () => {
	const store = createStore();
	const listener = mock();
	store.subscribe(listener);
	announceProvider(mockedProvider);
	const provider = store.getProviders()[0];
	expect(listener).toHaveBeenCalled();
	expect(provider).toEqual(mockedProvider);
});

it("should unsubscribe from store", () => {
	const store = createStore();
	const listener = mock();
	const unsub = store.subscribe(listener);
	unsub();
	announceProvider(mockedProvider);
	expect(listener).not.toHaveBeenCalled();
});

it("should clear providers list", () => {
	const store = createStore();
	const listener = mock();
	store.subscribe(listener);
	announceProvider(mockedProvider);
	store.clear();
	const providers = store.getProviders();
	expect(providers.length).toEqual(0);
});
