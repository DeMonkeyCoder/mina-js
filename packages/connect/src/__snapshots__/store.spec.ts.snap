// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`should initialize the store 1`] = `
{
  "_listeners": [Function: _listeners],
  "clear": [Function: clear],
  "destroy": [Function: destroy],
  "findProvider": [Function: findProvider],
  "getProviders": [Function: getProviders],
  "reset": [Function: reset],
  "subscribe": [Function: subscribe],
}
`;
